File: app/src/main/java/com/mohsenoid/rickandmorty/view/episode/list/EpisodeListPresenter.java
package com.mohsenoid.rickandmorty.view.episode.list;

import com.mohsenoid.rickandmorty.data.DataCallback;
import com.mohsenoid.rickandmorty.data.exception.EndOfListException;
import com.mohsenoid.rickandmorty.domain.Repository;
import com.mohsenoid.rickandmorty.model.EpisodeModel;
import com.mohsenoid.rickandmorty.util.config.ConfigProvider;

import java.util.List;

public class EpisodeListPresenter implements EpisodeListContract.Presenter {

    private final Repository repository;
    private final ConfigProvider configProvider;

    private EpisodeListContract.View view = null;

    private int page = 1;

    public EpisodeListPresenter(Repository repository, ConfigProvider configProvider) {
        this.repository = repository;
        this.configProvider = configProvider;
    }

    @Override
    public void bind(EpisodeListContract.View view) {
        this.view = view;
    }

    @Override
    public void unbind() {
        view = null;
    }

    @Override
    public void loadEpisodes() {
        if (view != null) view.showLoading();

        page = 1;
        queryEpisodes();
    }

    @Override
    public void loadMoreEpisodes(int page) {
        if (view != null) view.showLoadingMore();

        this.page = page;
        queryEpisodes();
    }

    private void queryEpisodes() {
        if (!configProvider.isOnline()) {
            if (view != null) view.showOfflineMessage(false);
        }

        repository.queryEpisodes(page, new DataCallback<List<EpisodeModel>>() {

            @Override
            public void onSuccess(List<EpisodeModel> episodes) {
                if (view != null) {
                    if (page == 1) {
                        view.setEpisodes(episodes);
                        view.hideLoading();
                    } else {
                        view.updateEpisodes(episodes);
                        view.hideLoadingMore();
                    }
                }
            }

            @Override
            public void onError(Exception exception) {
                if (view != null) {
                    view.hideLoading();
                    view.hideLoadingMore();
                }

                if (exception instanceof EndOfListException) {
                    if (view != null) view.reachedEndOfList();
                } else {
                    if (view != null) view.showMessage(exception.getMessage());
                }
            }
        });
    }
}

File: app/src/main/kotlin/com/mohsenoid/rickandmorty/view/episode/list/EpisodeListPresenter.kt
package com.mohsenoid.rickandmorty.view.episode.list

import com.mohsenoid.rickandmorty.domain.Repository
import com.mohsenoid.rickandmorty.domain.model.PageQueryResult
import com.mohsenoid.rickandmorty.util.config.ConfigProvider

class EpisodeListPresenter(
    private val repository: Repository,
    private val configProvider: ConfigProvider
) : EpisodeListContract.Presenter() {

    private var view: EpisodeListContract.View? = null

    private var page = 1

    override fun bind(view: EpisodeListContract.View) {
        this.view = view
    }

    override fun unbind() {
        view = null
    }

    override suspend fun loadEpisodes() {
        view?.showLoading()
        page = 1
        getEpisodes()
    }

    override suspend fun loadMoreEpisodes(page: Int) {
        view?.showLoadingMore()
        this.page = page
        getEpisodes()
    }

    private suspend fun getEpisodes() {
        if (!configProvider.isOnline()) {
            view?.showOfflineMessage(isCritical = false)
        }

        when (val result = repository.getEpisodes(page)) {
            is PageQueryResult.Successful -> {
                if (page == 1) {
                    view?.setEpisodes(result.data)
                } else {
                    view?.updateEpisodes(result.data)
                }
            }
            PageQueryResult.EndOfList -> view?.reachedEndOfList()
        }

        view?.hideLoading()
        view?.hideLoadingMore()
    }
}
